src/CoWSwapEthFlow.sol

```diff
@@ -0,0 +1,77 @@
+
+import "./libraries/EthFlowOrder.sol";
+import "./interfaces/ICoWSwapEthFlow.sol";
+import "./mixins/CoWSwapOnchainOrders.sol";
+
+/// @title CoW Swap ETH Flow
+/// @author CoW Swap Developers
+contract CoWSwapEthFlow is CoWSwapOnchainOrders, ICoWSwapEthFlow {
+    using EthFlowOrder for EthFlowOrder.Data;
+
+    /// @dev An order that is owned by this address is an order that has not yet been assigned.
+    address public constant NO_OWNER = address(0);
+
+    /// @dev The address of the contract representing the default native token in the current chain (e.g., WETH for
+    /// Ethereum mainnet).
+    IERC20 public immutable wrappedNativeToken;
+
+    /// @dev Each ETH flow order as described in [`EthFlowOrder.Data`] can be converted to a CoW Swap order. Distinct
+    /// CoW Swap orders have non-colliding order hashes. This mapping associates some extra data to a specific CoW Swap
+    /// order. This data is stored onchain and is used to verify the ownership and validity of an ETH flow order.
+    /// An ETH flow order can be settled onchain only if converting it to a CoW Swap order and hashing yields valid
+    /// onchain data.
+    mapping(bytes32 => EthFlowOrder.OnchainData) public orders;
+
+    /// @param _cowSwapSettlementContract The address of the CoW Swap settlement contract.
+    /// @param _wrappedNativeToken The address of the default native token in the current chain (e.g., WETH on mainnet).
+    constructor(address _cowSwapSettlementContract, IERC20 _wrappedNativeToken)
+        CoWSwapOnchainOrders(_cowSwapSettlementContract)
+    {
+        wrappedNativeToken = _wrappedNativeToken;
+    }
+
+    /// @dev Function that creates and broadcasts an ETH flow order that sells native ETH. The order is paid for when
+    /// the caller sends out the transaction. The caller takes ownership of the new order.
+    ///
+    /// @param order The data describing the order to be created.
+    /// @param quoteId The quote id obtained from the CoW Swap API to lock in the current price. It is included in the
+    /// extra information in the order creation event.
+    /// @return orderHash The hash of the CoW Swap order that is created to settle the new ETH order.
+    function createOrder(EthFlowOrder.Data calldata order, uint64 quoteId)
+        external
+        payable
+        returns (bytes32 orderHash)
+    {
+        if (msg.value != order.sellAmount) {
+            revert IncorrectEthAmount();
+        }
+
+        EthFlowOrder.OnchainData memory onchainData = EthFlowOrder.OnchainData(
+            msg.sender,
+            order.validTo
+        );
+
+        OnchainSignature memory signature = OnchainSignature(
+            OnchainSigningScheme.Eip1271,
+            abi.encodePacked(address(this))
+        );
+
+        // The data event field includes extra information needed to settle orders with the CoW Swap API.
+        bytes memory data = abi.encodePacked(onchainData.validTo, quoteId);
+
+        orderHash = broadcastOrder(
```

Comment:
I noticed that I misread the other PR, so commenting here.

Would it make sense for `broadcastOrder` to instead return an order UID instead of an order hash?

I guess there are additional gas costs you are trying to avoid?

Reply:
Yes, the extra allocation of a memory array is quite expensive (~530 gas for the emitter, ~660 gas for the order creation).

Here are the bench of the affected functions when returning the full orderUid. You can see the tested code with orderUid [here](https://github.com/cowprotocol/ethflowcontract/compare/bench-extra-cost-returning-orderuid).
```
 Function Name                                                                ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
 createOrder                                                                  ┆ 285             ┆ 26699 ┆ 36435  ┆ 36435 ┆ 6 
 createOrder                                                                  ┆ 285             ┆ 26146 ┆ 35773  ┆ 35773 ┆ 6
 Function Name                                                                                  ┆ min             ┆ avg   ┆ median ┆ max   ┆ # calls │
 broadcastOrderPublic                                                                           ┆ 10537           ┆ 10537 ┆ 10537  ┆ 10537 ┆ 2   
 broadcastOrderPublic                                                                           ┆ 9995            ┆ 9995  ┆ 9995   ┆ 9995  ┆ 2

```

Note that the tested code is a simplified version of the real code: this is because the actual owner address should be extracted from the signature and isn't available directly.

Reply:
Makes perfect sense. If an order UID is needed, then it can (as you mentioned in the other PR I think) be computed with `GPv2Order.packOrderUidParams`.

It might be good in the "general" `OnchainOrders` contract that we deploy to have a `broadcastOrderWithUid` so that you can write:
```sol
bytes memory orderUid = onchainOrders.broadcastOrderWithUid(...);
settlement.setPreSignature(orderUid);
```

But agree that for now, only the `broadcastOrder` returning the hash is needed.